### DISCUSSION TITLE
Typing is not correct for bracket-notation access for a Dataset variable

### DISCUSSION THREAD

**Discussion by eschalkargans** [pydata/xarray#8046] [Q&A] [answered]
Hi,

This is my first message here. I landed here following the "❓ Usage question" section in the [Issue creation interface](https://github.com/pydata/xarray/issues/new/choose). Indeed, this is more a question than an issue.

## Question

I noticed that type hinting in VSCode is not working for bracket-notation access for a Dataset variable. 

xarray version: `2023.6.0`

### Reproducible example

The following code-snippet is self contained. However, for type hinting, VSCode would be required. I assume any type-hinting mechanism would have the same issue.

```python
import xarray as xr 

ds = xr.Dataset({'da': xr.DataArray()})

# Dot notation
dot_notation = ds.da # Type hinting (in VSCode): `(type alias) dot_notation: Any`

# Bracket notation
bracket_notation = ds["da"] # Type hinting (in VSCode): `(variable) bracket_notation: DataArray` 
```

As you can see the type is widened to "Any", while "DataArray" would be the correct answer. Is it even possible to enforce type hinting on dot access notation? Indeed, we can dot-access other attributes of a DataArray, such as its methods, while bracket notation (`__getitem__`) is reserved for returning the Dataset variables?

I looked into the code for the `Dataset` class and indeed, its `__getitem__` method returns a `DataArray` when the given key is an Hashable (`str` in my case).

See https://github.com/pydata/xarray/blob/92c8b33eb464b09d6f8277265b16cae039ab57ee/xarray/core/dataset.py#L1506

So, does the xarray community recommends to favor bracket notation over dot notation to enforce type hinting in projects, and favor dot notation for shorter feedback loops (debugging, jupyter notebooks) where static typing is less important than type speed?

Thanks a lot!

PS: Regarding _I have done a search for similar discussions_, I looked for `typing` (https://github.com/pydata/xarray/discussions?discussions_q=typing) but did not find an issue exactly similar. Maybe https://github.com/pydata/xarray/discussions/5137 seems related.

**Comment by TomNicholas** [comment, marked as answer]
Hi, welcome! :wave: 

If I understand your question correctly, you want to know why the return type of the dot notation is `Any`. The answer is because the dot notation can also be used to access the `attributes` dictionary on xarray objects, which can store arbitary metadata. For example in this case the return type is a string:

```python
In [1]: import xarray as xr

In [2]: da = xr.DataArray(0, attrs={'t' : 'calendar'})

In [3]: da.t
Out[3]: 'calendar'
```

> So, does the xarray community recommends to favor bracket notation over dot notation to enforce type hinting in projects, and favor dot notation for shorter feedback loops (debugging, jupyter notebooks) where static typing is less important than type speed?

Exactly. At least that's what I do personally.

**Comment by eschalkargans** [comment]
Hello,

Thanks for your reply. 

What I did is replace the dot notations in my code by bracket notations to improve the typing coverage.
Also, it reminds me of the pitfall with dot notation related to reassignment, where bracket notation is mandatory.

```python
da.x = something # won't work
da["x"] = something # works
```

So overall it seems the bracket notation is more "robust"!
